### PR TITLE
Store AWS region in AwsStorageConfigurationInfo

### LINKED ISSUE #342: [FEATURE REQUEST] support for returning client region in the loadTable response for S3
### Is your feature request related to a problem? Please describe.

AWS s3 by default fails when making cross region calls. Latest version of AWS s3 sdk support crossRegionAccess but Iceberg doesn't yet support this [functionality](https://github.com/apache/iceberg/pull/9804). 
Any cross region calls are failing 

### Describe the solution you'd like

It would be great to specify the client.region as part of loadTableResponse in `config` block along with Credentials. This is property defined by RESTCatalog openAPI [spec](https://github.com/apache/iceberg/blob/b1d38b3cac326da296b770c8ab9219bc4f791666/open-api/rest-catalog-open-api.yaml#L3127)

### Describe alternatives you've considered

Using the crossRegionAccess property. Maybe once Iceberg add support, it would be great to add this property to the spec too

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by eric-maynard**
# Description

This adds support for a new property, `region` for AWS storage configurations.

Fixes #342

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation update
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

I'm able to create catalogs and add the `region` property to the StorageConfigInfo:
```
LD4RTJ0HY9:polaris emaynard$ curl -X POST http://localhost:8181/api/management/v1/catalogs \
> -H "Authorization: Bearer principal:root;realm:default-realm" \
> -H "Content-Type: application/json" \
> -d '{
>   "catalog": {
>     "type": "INTERNAL",
>     "name": "example_catalog",
>     "properties": {
>       "default-base-location": "s3://your-bucket/catalog-location/"
>     },
>     "storageConfigInfo": {
>       "storageType": "S3",
>       "roleArn": "arn:aws:iam::012345678901:role/jdoe",
>       "region": "us-east-2"
>     }
>   }
> }'
LD4RTJ0HY9:polaris emaynard$ curl -X GET http://localhost:8181/api/management/v1/catalogs/example_catalog \
> -H "Authorization: Bearer principal:root;realm:default-realm" \
> -H "Content-Type: application/json" | jq
{
  "type": "INTERNAL",
  "name": "example_catalog",
  "properties": {
    "default-base-location": "s3://your-bucket/catalog-location/"
  },
  "createTimestamp": 1731818113312,
  "lastUpdateTimestamp": 1731818113312,
  "entityVersion": 1,
  "storageConfigInfo": {
    "storageType": "S3",
    "roleArn": "arn:aws:iam::012345678901:role/jdoe",
    "externalId": null,
    "userArn": null,
    "region": "us-east-2",
    "allowedLocations": [
      "s3://your-bucket/catalog-location/"
    ]
  }
}

```

### CODE REVIEW

**Review by singhpk234 — COMMENTED** [review]
[doubt] This region specified will be region of the polaris catalog ? Do we plan to support federation of catalogs in future if yes, the should this be the region of polaris catalog or the region of federated catalog ? 

for ex: 
can this a possible case in coming future ? 
Polaris  -> (Federated to) -> Glue 
(region A)      ------ > (region B)

**Comment by eric-maynard** [comment]
Hey @singhpk234, the idea is that you can associate a region with a storage configuration. As you pointed out, it is not clear how this will work with catalog federation (cc @dennishuo). But I think it is also unclear how storage configurations will work with federated catalogs more generally -- for example, a single role ARN may not be valid for the entire federated catalog. So this is something our design for federation must address.

At the very least, we have discussed allowing storage configurations to be defined on a level more granular than the catalog (e.g. at the table or namespace level).

Maybe @munendrasn, the filer of #342, can also help provide some additional context here. For my part I am curious if there's a particular test case we can add here to make sure the issue reported in #342 is fully addressed.

**Comment by eric-maynard** [comment]
Hi @munendrasn I see -- do the current changes here work for your use case then? `client.region` should be specified in the credentials map so long as it's set for the table's storage configuration.

**Comment by munendrasn** [comment]
@eric-maynard 
As long as `client.region` is set, it would the reported issue. Skimming through the code, I couldn't find `client.region` being returned as part of credentialMap. would it be possible to include a test case for the same?

**Comment by eric-maynard** [comment]
Hi @munendrasn, please see the newly-added test. `client.region` should be present in the credentials map when a region is attached to the storage config.

File: polaris-core/src/main/java/org/apache/polaris/core/storage/aws/AwsCredentialsStorageIntegration.java [hunk 1]
```diff
@@ -76,6 +76,7 @@ public EnumMap<PolarisCredentialProperty, String> getSubscopedCreds(
     credentialMap.put(
         PolarisCredentialProperty.AWS_SECRET_KEY, response.credentials().secretAccessKey());
     credentialMap.put(PolarisCredentialProperty.AWS_TOKEN, response.credentials().sessionToken());
+    credentialMap.put(PolarisCredentialProperty.CLIENT_REGION, storageConfig.getRegion());
```
Comment (munendrasn) [inline polaris-core/src/main/java/org/apache/polaris/core/storage/aws/AwsCredentialsStorageIntegration.java:79]: nit: should we add `client.region` only if region is set?

**Review by munendrasn — APPROVED** [review]
lgtm

**Review by munendrasn — APPROVED** [review]

File: polaris-core/src/main/java/org/apache/polaris/core/storage/aws/AwsStorageConfigurationInfo.java [hunk 2]
```diff
@@ -49,23 +49,30 @@ public class AwsStorageConfigurationInfo extends PolarisStorageConfigurationInfo
...
+
   @JsonCreator
   public AwsStorageConfigurationInfo(
       @JsonProperty(value = "storageType", required = true) @NotNull StorageType storageType,
       @JsonProperty(value = "allowedLocations", required = true) @NotNull
           List<String> allowedLocations,
-      @JsonProperty(value = "roleARN", required = true) @NotNull String roleARN) {
-    this(storageType, allowedLocations, roleARN, null);
+      @JsonProperty(value = "roleARN", required = true) @NotNull String roleARN,
+      @JsonProperty(value = "region", required = false) @NotNull String region) {
```
Comment (adutra) [inline polaris-core/src/main/java/org/apache/polaris/core/storage/aws/AwsStorageConfigurationInfo.java:62]: Why `@NotNull` here while the property is nullable?

Comment (eric-maynard) [reply polaris-core/src/main/java/org/apache/polaris/core/storage/aws/AwsStorageConfigurationInfo.java]: Good catch, I thought it had to be `NotNull` for the `JsonCreator` but in fact that's not true. Fixed.

**Review by adutra — APPROVED** [review]

**Comment by adutra** [comment]
@eric-maynard let's merge this one, could you please resolve the conflicts?

**Comment by eric-maynard** [comment]
Hey @adutra, the conflicts are resolved -- when you get a moment, could you re-approve? Thanks!

**Review by adutra — APPROVED** [review]

**Pull request merged**